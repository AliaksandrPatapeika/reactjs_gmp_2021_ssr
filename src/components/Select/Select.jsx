import {filter, includes} from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import ReactSelect from 'react-select';

import ClearIndicator from './components/ClearIndicator';
import DropdownIndicator from './components/DropdownIndicator';
import MultiValue from './components/MultiValue';
import Option from './components/Option';
import RequiredSelect from './components/RequiredSelect';
import ValueContainer from './components/ValueContainer';

/* https://github.com/JedWatson/react-select/issues/3543#issuecomment-565733597 */
const Select = (props) => {
  const {isMulti, allowSelectAll} = props;

  if (isMulti && allowSelectAll) {
    return (
      <RequiredSelect
        // PATTERN: Spreading props on DOM elements
        {...props}
        SelectComponent={ReactSelect}
        options={[props.allOption, ...props.options]}
        components={{
				  ClearIndicator,
				  DropdownIndicator,
				  MultiValue,
				  Option,
				  ValueContainer
        }}
        onChange={(selected, event) => {
				  if (selected !== null && selected.length > 0) {
				    if (selected[selected.length - 1].value === props.allOption.value) {
				      return props.onChange([props.allOption, ...props.options]);
				    }
				    let result = [];

				    if (selected.length === props.options.length) {
				      if (includes(selected, props.allOption)) {
				        result = filter(selected, (option) => option.value !== props.allOption.value);
				      } else if (event.action === 'select-option') {
				        result = [props.allOption, ...props.options];
				      }
				      return props.onChange(result);
				    }
				  }

				  return props.onChange(selected);
        }}
      />
    );
  }

  return (
    <ReactSelect
      {...props}
      components={{
			  DropdownIndicator
      }}
    />
  );
};

Select.propTypes = {
  options: PropTypes.instanceOf(Array).isRequired,
  value: PropTypes.instanceOf(Object),
  onChange: PropTypes.func.isRequired,
  isMulti: PropTypes.bool,
  allowSelectAll: PropTypes.bool,
  allSelectedLabel: PropTypes.string,
  allOption: PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired
  })
};

Select.defaultProps = {
  allSelectedLabel: 'All is selected',
  isMulti: false,
  allowSelectAll: false,
  value: null,
  allOption: {
    label: 'Select all',
    value: '*'
  }
};

export default Select;
