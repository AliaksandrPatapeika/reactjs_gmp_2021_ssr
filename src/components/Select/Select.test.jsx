import {noop} from 'lodash';
import React from 'react';
import renderer from 'react-test-renderer';

import Select from './Select';

describe('Select', () => {
  test('renders correctly', () => {
    const select = renderer
      .create(<Select onChange={noop} options={[]} />)
      .toJSON();

    expect(select).toMatchSnapshot();
  });
});
