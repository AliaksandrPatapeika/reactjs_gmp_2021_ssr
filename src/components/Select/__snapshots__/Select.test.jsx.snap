// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<div
  className=" css-2b097c-container"
  onKeyDown={[Function]}
>
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    className="css-1f43avz-a11yText-A11yText"
  />
  <div
    className=" css-yk16xz-control"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className=" css-g1d714-ValueContainer"
    >
      <div
        className=" css-1wa3eu0-placeholder"
      >
        Select...
      </div>
      <div
        className="css-b8ldur-Input"
      >
        <div
          className=""
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "label": "input",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className=" css-1hb7zxy-IndicatorsContainer"
    >
      <span
        className=" css-1okebmr-indicatorSeparator"
      />
      <div
        aria-hidden="true"
        className=" css-tlfecz-indicatorContainer"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-caret-down"
        />
      </div>
    </div>
  </div>
</div>
`;
